GH-631: Review 2

utkarshg6 opened the pull request.

czarte committed:
added dbip_download github workflow (#637)

* added dbip_download github workflow

* change to run every day on midnight for QA

czarte committed:
Master gh 580 b dbip download workflow (#639)

* added dbip_download github workflow

* change to run every day on midnight for QA

* necesary changes to be able build dbip_country.rs on master

* -p for generating ip_country/src

czarte committed:
Master gh 580 b dbip download workflow (#641)

* added dbip_download github workflow

* change to run every day on midnight for QA

* necesary changes to be able build dbip_country.rs on master

* -p for generating ip_country/src

* todo in dbip_download.yaml

* condition for new branch generated-source

* fixing condition for generated-source branch

czarte committed:
Master gh 580 b dbip download workflow (#643)

czarte committed:
Master gh 580 b dbip download workflow (#644)

* added dbip_download github workflow

* change to run every day on midnight for QA

* necesary changes to be able build dbip_country.rs on master

* -p for generating ip_country/src

* todo in dbip_download.yaml

* condition for new branch generated-source

* fixing condition for generated-source branch

* fix generated-source branch name in condition

* increse standard timeout milis const

* manual trigger and debugging flags to runs

* use variable to store git ls-remote command

utkarshg6 committed:
GH-631: add review changes for node/src/database/db_initializer.rs

czarte committed:
Master gh 580 b dbip download workflow (#647)

* added dbip_download github workflow

* change to run every day on midnight for QA

* necesary changes to be able build dbip_country.rs on master

* -p for generating ip_country/src

* todo in dbip_download.yaml

* condition for new branch generated-source

* fixing condition for generated-source branch

* fix generated-source branch name in condition

* increse standard timeout milis const

* manual trigger and debugging flags to runs

* use variable to store git ls-remote command

* add git fetch before checkout

utkarshg6 committed:
GH-631: some more review changes

utkarshg6 committed:
GH-631: some more review changes

czarte committed:
Master gh 580 b dbip download workflow (#648)

* added dbip_download github workflow

* change to run every day on midnight for QA

* necesary changes to be able build dbip_country.rs on master

* -p for generating ip_country/src

* todo in dbip_download.yaml

* condition for new branch generated-source

* fixing condition for generated-source branch

* fix generated-source branch name in condition

* increse standard timeout milis const

* manual trigger and debugging flags to runs

* use variable to store git ls-remote command

* add git fetch before checkout

* fix workflow for nonexistent branch

* fixing edgecases on nonexistent branch and no git changes

* change neverssl to different web

utkarshg6 committed:
GH-631: change checked to a boolean

utkarshg6 committed:
GH-631: add more changes

czarte committed:
Master gh 580 b dbip download workflow (#649)

utkarshg6 committed:
GH-631: keep pushing more review changes

utkarshg6 committed:
GH-631: display more info in the error statements

utkarshg6 committed:
GH-631: and even more changes

utkarshg6 committed:
GH-631: change checked to rechecked

utkarshg6 committed:
GH-631: remove blank line

utkarshg6 commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -68,7 +75,10 @@ pub trait FailedPayableDao {
-    fn update_tx_failures(&self, hash_map: &FailureUpdates) -> Result<(), FailedPayableDaoError>;
+    fn update_recheck_status(
+        &self,
+        hash_map: &HashSet<TxHash>,
```
This should be renamed to `hash_set`.

utkarshg6 commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -528,64 +566,106 @@ mod tests {
+            .hash(make_tx_hash(5))
+            .reason(FailureReason::PendingTooLong)
+            .rechecked(false) // This one is rechecked
+            .timestamp(now - 7200) // 2 hours ago
```
This should be rechecked and within the time boundary, maybe pick 40 minutes.

utkarshg6 commented on node/src/accountant/db_access_objects/sent_payable_dao.rs:
```diff
@@ -627,6 +638,8 @@ mod tests {
         let tx1 = TxBuilder::default().hash(make_tx_hash(1)).build();
         let tx2 = TxBuilder::default().hash(make_tx_hash(2)).build();
+        let tx1_prior_block_opt = tx1.block_opt.clone();
+        let tx2_prior_block_opt = tx2.block_opt.clone();
```
Instead of `clone()`, use `is_some()`.

utkarshg6 committed:
GH-631: minor self-review changes

bertllll commented on node/src/accountant/db_access_objects/test_utils.rs:
```diff
@@ -94,7 +108,22 @@ impl FailedTxBuilder {
     }
 }
+
+pub fn make_read_only_db_connection(home_dir: PathBuf) -> ConnectionWrapperReal {
```
I just hope we hadn't had this function already. Could you just somehow quick-check? By a string search with some phrases that come on your mind as possible to be used in such context. Otherwise all good. Great job.

utkarshg6 replied:
I found `trick_rusqlite_with_read_only_conn`, which creates a read-only database connection but also creates a table.

bertllll commented on node/src/database/db_initializer.rs:
```diff
@@ -727,7 +727,8 @@ mod tests {
             .prepare("select name, value, encrypted from config")
             .unwrap();
-        let _ = stmt.query_map([], |_| Ok(42)).unwrap();
+        // TODO: The next line's execution returns Err(ExecuteReturnedResults). Don't know why.
```
I think execute in the rusqlite library is meant to return only descriptive numbers not any data itself. 
If you're employing an SQL statement which is supposed to return a value (the one stored in the database column), you need to use `query`-like methods. For example `query_map`.

utkarshg6 replied:
Ok, then I'll just revert it back to the original statement.

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -47,18 +47,25 @@ pub struct FailedTx {
 
 pub enum FailureRetrieveCondition {
-    ByReason(FailureReason),
+    UncheckedPendingTooLong(u32), // u32 represents seconds ago
```
Why do we need to work with seconds? :thinking: 

That's weird, I think we only consider a single "pending payable interval" the boundary of the lapse. It should technically the interval configured in the body of the Accountant (either given by the user or defaulted). However, given the strangeness of the real world conditions, this number of seconds will always differ. (It depends on the message queue). I just wonder if we do want to know...

Maybe yes, but only for some kind of talkative logging

utkarshg6 replied:
Ok, got it! I think I should eliminate it.

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -47,18 +47,25 @@ pub struct FailedTx {
+                write!(
+                    f,
+                    "WHERE reason = 'PendingTooLong' AND rechecked = 0 \
+                     AND timestamp >= {} \
```
Why do we have to care about the time?

You think we cannot rely on the `rechecked` column? We had some discussion and I think I give you enough trust in what I was saying that you now follow...
Let's keep time out of this whole business. The line 64 is all we need.
 
In every scan cycle we should always update the latest PendingTooLong errors from 0 to 1. Therefore we always can rely on the 0. If we ask about the last, we simply ask about these 0s.  

(Note: I mightn't have explained enough but this function isn't used regularly. I suppose we will call it only once in the life span of the Node. Otherwise, we will only keep writing these records in and also mirroring those in the runtime memory just to avoid the need of re-reading them like this so often...it'll be much easier to work with the hashmap. But if the Node panics, the hashmap will always be backed up in the persistent db).

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -68,7 +75,10 @@ pub trait FailedPayableDao {
     fn insert_new_records(&self, txs: &[FailedTx]) -> Result<(), FailedPayableDaoError>;
     fn retrieve_txs(&self, condition: Option<FailureRetrieveCondition>) -> Vec<FailedTx>;
-    fn update_tx_failures(&self, hash_map: &FailureUpdates) -> Result<(), FailedPayableDaoError>;
+    fn update_recheck_status(
```
This should probably be either "update_status_rechecked" or simply "mark_as_rechecked".

bertllll replied:
Also, in relation to the confusions regarding the time test upon the retrieval of the pending payable, I wanna request you to change this method so that it actually always flip the flag of EVERY record conditioned as 
`reason = PendingTooLong AND rechecked = 0`.

Please raise an argument, if you can think of a good one, why we shouldn't do it. I myself reason about that 
the ultimate state which we want to achieve is supposed to be consistent with this idea I put forth. We actually EXPECT every 0 to change to 1 always just after a single cycle finishes, none of these 0 should stay longer than that to be resolved sometime later. That's not what we want. It definitely should be framed only by the stretch of one cycle. When we fully adopt this thinking we can perhaps agree on the stricter, but simpler at the same time, conditions I suggest. 

If you worry what if we are missing something, because I am partially hit by these questions too, you may free your spirit. I have given it some more time and assured myself that eventually, if this absoluteness (that means the use of no time parameter in the query) did not prove true, our whole mechanism will break down. And that will give us the answer. People with this issue would report the payments not working. In their logs, there would be an evidence of an endless loop of attempts, and we'd find out, checking the blockchain that their further transactions were always rejected due to the use of the wrong nonce.  
We can fix it then accordingly.     
     
Notice that this means that this method would need no additional arguments! :wink:

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -209,7 +231,8 @@ impl FailedPayableDao for FailedPayableDaoReal<'_> {
             let reason =
                 FailureReason::from_str(&reason_str).expect("Failed to parse FailureReason");
-            let checked: u8 = row.get(9).expectv("checked");
+            let checked_integer: u8 = row.get(9).expectv("rechecked");
```
Maybe "rechecked_as_int"?

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -362,15 +386,29 @@ mod tests {
+        let tx2 = FailedTxBuilder::default()
+            .hash(hash)
+            .rechecked(true)
+            .build();
```
Again... this recheck == true isn't technically possible for "insert_new_records". 
It irks me. :scream:

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -386,7 +424,10 @@ mod tests {
-        let tx2 = FailedTxBuilder::default().hash(hash).checked(1).build();
+        let tx2 = FailedTxBuilder::default()
+            .hash(hash)
+            .rechecked(true)
```
Same here and in so many other places :cry:

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -528,64 +566,106 @@ mod tests {
-        let result = subject.retrieve_txs(Some(ByReason(PendingTooLong)));
-
-        assert_eq!(result, vec![tx1, tx2]);
+        // Retrieve unchecked PendingTooLong transactions from the last hour
```
I don't really know what the time has to do with this. 

Could you please explain?

Edit: There is already a comment from me at the code this touches in particular. No need to reply to this question...

bertllll reviewed (commented):
continue in node/src/accountant/db_access_objects/failed_payable_dao.rs
`can_retrieve_unchecked_pending_too_long_txs`

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -68,7 +75,10 @@ pub trait FailedPayableDao {
+        hash_set: &HashSet<TxHash>,
+    ) -> Result<(), FailedPayableDaoError>;
     fn delete_records(&self, hashes: &HashSet<TxHash>) -> Result<(), FailedPayableDaoError>;
 }
```
I can't comment on the `SentPayableDao` trait in this iteration, but I need to add an important request on you. 

We have discussed it in the chat roughly already. 

The thing is that we've got a void for the design of the procedure when we uncover that a transaction tracked as failed for pending too long was tagged so unrightfully. 

We have defined correctly that it should be taken out from the failed payable table. Now I think that actually is just the least important part. 

With regard to a possible panic-caused data loss and the resulted database inconsistent state dragging this system into the lands of no recovery, it is important to think about the steps so that the database must stay sane and serve its purpose even after a sudden death of the application. We need to consider the vulnerability at every moment separately. 
    
The steps should be:
a) rewrite the newer (successor) record dwelling recently in sent_payable of the nonce N with the data of the older transaction with the same nonce, now read from the failed_payable table. I say rewrite because I want to avoid the db state inconsistency precisely.
That means, I'd like to have a single method to call, which will internally perform: the old row deletion + insertion of the new row, but committed no earlier than both of them done. I say we should use an explicit rusqlite tx, that gives you the control over when it is committed. 
However, you may be able to work it around by writing an UPDATE sql which will update the whole row though (every valuei in the row.) I wonder if that is considered a hack in the community of the data-engineers? (According to the ChatGPT, it isn't a nonsense. I'd then highly recommend that as it is much simpler to implement)  
b) delete the record from failed_payble 

Maybe 
`fn replace_record()`
or
`fn rewrite_whole_record()`

utkarshg6 replied:
I was considering whether this function should be placed outside of these two DAOs, as it doesn't fully belong to either one. What are your thoughts?

bertllll replied:
Nope, I believe it fits perfectly fine to the SentPayableDAO.

Please write it as a simple update statement which changes every value in the respective row. It's actually an idiomatic approach.

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -528,64 +566,106 @@ mod tests {
+        let hash_set = HashSet::from([tx2.hash, tx3.hash]);
 
-        let result = subject.update_tx_failures(&hash_map);
+        let result = subject.update_recheck_status(&hash_set);
```
As stated by the trait definition, this method is suspected of a full loss of args, because it probably won't need them anymore.

czarte committed:
Gh v2 580b czarte (#629)

* test for exit location in large patch

* implementing exit_location

* ExitLocation struct and its members to messages.rs, set_exit_location_command to masq, implemented From(String, usize) for CountryCodes

* Gh 784 review four (#521)

* continue of code review changes

* test for panic in handle_retrieve_transactions

* fixing more comments from review

* removing TODO

* implementing review, added two new tests for ip_country

* removing unnecesary country_code from node_descriptor

* added cfg(test) and cfg(not(test)) for static ref COUNTRY_CODE_FINDER, created test_dbip_country.rs

* change country_code member of NodeRecordInner_0v1 to country_code_opt

* simplifying tests in node_record.rs and neighborhood_database.rs

* resolving review comments

* final changes in gossip_acceptor from review

* continue of implementing review comments

* fix test return value

* added ip_country to ci/all.sh

* cherry-pick from GH-784-fix-for-Vojta and implementing iterator for deserializer

* try to create test for fmt

* tests for print ip v4 and v6 vec[u8] in country_block_serde.rs

* ci for ip_country + resolved conflicts of clippy

* masq_lib dependencies solved

* dependency masq_lib for ip_country in node workspace

* remove commented out code + fix condition for non empty bit_queue

* formatting

* 5 seconds for wait_for_log in requested_chain_meets_different_db_chain_and_panics_integration

* fixing impl for AGR in multinode tests

* formatting

* fix nonexistent NodeAddr in multinode tests for AGR

* change single match to if

* try to fix build in actions for MacOS

* try to find out disk space on MacOS container in Actions

* add diagnostics commands on failure

* check diskspace before ci/all.sh workflow runs

* diagnose mounted disks

* analyze /var/ diskspace

* remove advanced diagnosticst for disk space

* switch macos to version 13

* diagnostics for /Users/runner/work/Node/Node/node/target/release

* fasten diagnostics

* fasten diagnostics fix

* fasten diagnostics back

* GH-784-review-two-state: fixes that could work on Actions, appeasing it

* fixing review

* implementing CountryBlock count in generated rs file, and infrastructure

* fixing ip_country tests

* fix import of ip_country to multinode_integration_tests

* formatting

* fix the reference for COUNTRY_CODE_FINDER

* implemenitng initalization checker for COUNTRY_CODE_FINDER

* CountryCodeFinder initializer

* fixed comments from review

* removed custom PartialEq for NodeLocation

* get back with COUNTRY_CODE_FINDER

---------

Co-authored-by: Bert <Bert@Bert.com>

* implemented country-codes and fallback-routing into MessageBody for Node from the UI

* completed test for c
ommand for exit location, add test for neighborhood actor to test recieved message

* implemented handler for exit_location_opt, rename exit_loctation to exit_location_opt, cleanup persistent configuration, test finished for set_exit_location message handler

* fmt for HashSet<ExitLocation>, finished the assertions in test exit_location_can_be_changed_using_exit_location_msg

* cleanup, test for fallback-routing: false, logger for fallback-routing and implementation for handle from message in Neighborhood

* change exit_locations_opt from HashSet to HashMap

* clearing bootstrapper config

* country_undesidability to NodeRecordMetadata

* added member for NodeRecordMetadata and processing for country_undesirability in message handler for exit location

* added assertion message for assert on db change in actor

* refacotring neighborhood_test_utils: extracted make_segmented_ip, and make_segments for usage in tests, add comments to assertions in handle_exit_location test, test for nodes_mut

* added unset exit_location logic with test

* change name of test

* fiixing tests for set_exit_location_command, fixing log assertion for exit_location handler in Neighborhood

* added assertion on neighborhood to exit_location tests

* pull request branch changed to GH-784

* formatting

* fixing comments from review

* polishing helps

* addressing comments from review

* polishing helps for exit-location

* add use super::*; to tests use

* add cfg(test) for tests in set_exit_location_command

* fixing clippy complains

* fix multinode neighborhood constructor

* rename set_exit_location_command.rs to exit_location_command.rs

* add protection for empty Neighborhood DB

* change protection for empty Neighborhood

* fixed shared schema, added exit location to --help of masq

* set version for sccache

* add lifetime for daemon_shared_app fn return

* fixing log assertion in tests, add db to exit location test, fix app lifetime for daemon help

* introduce test utilities

* remove unused import - localhost linux compile and passes all ip_country tests

* addressing comments from review

* addressing comments from review, refactoring Neighborhood for exit-location, introduce enum for ExitPreference, fixed Display for ExitLocationSet

* fix tests for separate_u64_values

* add Eq to ExitPreference

* rearranged test_utilities, fix to exit_location_handler and its tests

* fixing ByteArrayWriter and Reader for masq

* added more nodes to exit_location test with fallback to test UNREACHABLE_COUNTRY_PENALTY

* move test_utilities to dev-dependencies

* test_utilities from dev to production dependencies in masq_lib

* formatting help

* formatting help

* review 3 - refactoring message handler for exit location, and few minor changes

* country codes for NodeRecords in tests Neighborhood DBs, constants for assertion and better error messages, other minor changes

* formatting

* fix todo comment, and formatting

* exit-loaction data structures moved into the ExitTools, cloned into NeighborhoodMetadata and sent to handler, modified test for gossip handler to ensure new node has modified undesirability

* implemented adding country_undesirability for introducer and debutant

* formatting and remove unnecesary let

* fallback routing assertion change

* formatting

* rename ExitTools

* fix request for neverssl.com to impersonate user-agent

* changes in test for country_code_route, remove country_code_opt from routing engine, replaced by UserExitPreferences, implemented filtr for country_code in routing

* working exit country selection wihtout fallback, introduces new structure db_countries and find_exit_location fn from GH-469

* start of implementing find_exit_location

* changes to create db_countries

* completed process for db_countries and routign_engine for fallback_routing

* handling missing locations in NeighborhoodDB

* introduced Error responose for ExitLocation handler, adjusted tests, investigate find_exit_location function and routng_engine

* modified routing engine to harvest more exit nodes from Neihgborhood DB

* implemented recursion for research neighborhood, adjusted tests, missing_countries moved to extract_exit_locations_from_message

* fix research_neighborhood in routing_engine, new recursion call for research_neighborhood, recursion guts moved to outbound function

* formatting and compiler complains resolved

* test for db_countries setting on handle gossip

* completed test for newly introduced exit node, added assertions, works

* test for removing exit location of last node with particular country code from db_countries, TODO created

* formatting

* removing comments, introducing TODOs

* unit test for init_db_countries(), removed duplicit test for handle gossip

* added --show-countries for exit-location subcommand for testing purposes

* first cherry pick from amoy commit b2a3ebf85199b3c1a6620a9047aaab41090b9b52

* added amoy network

* formatting

* amoy fixing

* fixing routing engine for direction back in route, and fallback-routing condition

* create prefix with capacity

* back to previous version of routing_engine

* fix dotgraph

* allow dead code for unused service functions for new multinode integration test of exit_location

* fix tilde for windows - cherry pick from master

* formatting

* remove unused imports

* fix tilde in config-file path from master

* replace neverssl.com for testingmcafeesites.com in multiple_streans_zero_hop

* formatting

* remove slash from url in test to pass

* multinode test with processing functions

* remove unused imports and questionmark

* connect last node to germany network

* interconnect_world_network public

* remove networks before creation

* add creation of integration_net to test

* remove world network from last node

* fix first node

* network into docker run

* add clone, debug, partialeq to NetworkPack

* fix arguments count to do_docker_run

* fix do_prepared_docker_run

* solving temporary issue

* solving dereference

* solving free while use issue

* resolve lifetime issues

* resolve lifetime issues do_docker_run

* remove network connect duplication

* fix ip start index for nodes in loop

* fix ip start index for nodes in loop cz

* fix ip start in loop

* fix ip start in loop de

* multinode integration test with multiple real networks, selecting exit-location country, and retrieving web result

* country_code_exit_location_routing - final stage with multiple networks

* multinode test http_end_to_end_routing_test_with_exit_locationL

* rebuilding exit-location command response

* introduced UiSetExitLocationResponse and processing it in masq CLI, fixed exit-location handler and messages

* multinode test and refactoring masq command request and response

* resolving self-review: updated copyright to right year, removing obsolete TODOs, added test for constants, corrected output to stdout in exit_location_command

* added assertions Response and udated Response processing exit-location-command tests

* remove unnecessary Docker components for exit_location test, fixing self review commnets, new TODO in neighborhood/mod.rs

* resolved handle_new_public_ip test, and half way to resolve handle_gossip test and update db_countries on DB change

* test for init_db_countries in handle_agrs, condition for one-hop in handle_agrs Reply, remove target_country_opt

* formatting

* fix exit-location cmd assertion

* multinode test from master

* make_node_record do not add country if show_ip is false

* ok_responose partialy from master for MockBlockchainClientServer

* resolved last comments

* remove data from dbip_country.rs - prepared for actions build

* working on exit_location_command

* addressing issues from review

* continue addressing review comments

* Added DBIP file generation

* ci: improve GitHub Actions workflow - Replace s3-sync-action with AWS CLI, update credentials handling, improve cache invalidation

* refactor: improve path resolution and code quality in generate_dbip

* ci: remove unnecessary dependencies and update generate dbip step

* chore: remove unused generate_dbip_country.sh script

* docs: update download script message to reference cargo command

* refactor: Improve generate_dbip.rs organization and efficiency

* fix: Update dbip_country.rs paths in CI workflow to use temp directory

* fix: Update verification pattern for generated dbip file

* ci: revert platform changes to legacy

* ci: update windows specific file move to use bash

* ci: corrected verification step

* more addressed comments from review

* test for min_hops_change affects exit_countries and enrich_exit_countries in extract_exit_locations_from_message to be implemented in handle_config_change_msg

* new machinery for min-hop affects exit location settings

* handle_config_change_msg debugging

* finalize min_hops_change_affects_db_countries test

* rename test

* refactoring neighborhood/mod.rs, moving AccessibleGossipRecord to gossip.rs, formatting

* Room for an MMDB parser

* Generating countries, rudimentary

* parse() should produce a Countries object, instead of Vec<(String, String)>. Gonna do that.

* fix db to dot grap test

* Interim commit

* addressing more comments from review

* addressing minor changes from review

* addressing review comments

* resolving comments from review

* Must make tests self-adapting

* No MMDB yet, but tests pass

* addressing comments from review

* UI README updated (uncompleted), comment explaining country_undesirability in Metadata, addressing gomments from review

* About to try bringing in an actual DBIP file

* Happy-path test passing

* formatting for ExitLovationSet

* setup_reporter_shared_app and wait_for_masq_node_ends

* CountryCodes to CountryBlocks, fix formatting of ExitLocationSet in tests, remove reference from supporting functions, already reference comming in

* resolving unimplemented Copy trait in Process

* new clap test providing_show_countries_with_other_argument_fails

* format

* MMDB parser seems to be finished

* fallback routing test change

* country codes to nodes names in tests

* description to exit-location response

* removing comments, adding country codes to node names in tests

* Now MMDB parser isn't as wasteful

* Removed dead data

* Tests all pass

* Merge mostly complete; lots of tests pass. Maybe all of them?

* One test failing. Can't figure why.

* All masq tests passing now

* Multinode tests pass; one bit of code cleanup remains

* All tests should be passing

* Removed dead code

* Little surgery on modules and imls

* Updated the CSV country list from MMDB

* Started GitHub Actions file for monthly downloads

* conditions for failing download and gargo run, to not continue, file names in generate and commit

* Going to try out the dbip_download workflow

* Temporary on:

* addressing review 2 comments

* addressing comments from review2

* fixing line length in user interface interface md

* tests to remove unexistent countries from user_exit_preferences.locations_opt

* remove tests and commnet of filtering out unexistent country_codes

* comment about the locations_opt

* make_node_record_cc and adjusting tests, get rid of free world bit

* fixes country_code on tests, introduced make_node_record_cc

* addressing commnets from review3

* remove pub from CountryCodeFinder member, and remove commented out code

* utilization of strings in clap test for show-countries

* last review with master merged in

* forgotten change

* generate dbip_country.rs workflow and checkout dbip_country.rs from separate branch in ci-matric workflow

* resolved most failing tests, now failing debut_with_node_addr_not_accepting_connections_is_rejected for no reason, wierd behaviour of qualifies

* fixing failing tests

* comment out fetch dbip, and remove unwanted action build dbip

* try to fix actions

* all countries in test dbip_country.rs

* removing misstake

* added checkout of read dbip_country.rs from DBIP-Experiment

* new workflow: in local we work with reduced dbip, but on ci/all.sh we checking out full version from git, same in actions, remove PARENT_DIR variable for Jenkins

* testing git pre-commit hook

* update pre-commit hook

* remove checkout of dbip_country.rs from ci-matrix, it will be handled by ci/all.sh

* testing pre-commit hooks changes

* testing commit hook

* testing pre commit hook

* testing pre commit hook

* final polishing of pre-commit hook

* addressed comments from Review

* formatting

* addressing review comments, plus fix double entry of ip_country in modules.xml

* Enabled IDE tests in ip_country

* Fixed module-definition problem, addressed two review comments

* Weird stuff with TryFrom and From

* addressing review comments

* fix ci.all condition for chekout dbip

* remove comments and exit in ci/all.sh

* generated-source from Node repo

* dbip download workflow for master

* adjust cron shedule for dbip download

* final touch for dbip workflow

* git pull before ci/all.sh

---------

Co-authored-by: Bert <Bert@Bert.com>
Co-authored-by: FinsaasGH <89403560+FinsaasGH@users.noreply.github.com>
Co-authored-by: FinsaasGH <finsaas@protonmail.com>
Co-authored-by: Dan Wiebe <dnwiebe@gmail.com>

utkarshg6 committed:
GH-631: add another review changes

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -139,6 +132,13 @@ impl FailedPayableDao for FailedPayableDaoReal<'_> {
+                txs
+            )));
+        }
+
```
Thanks for this :hugs:

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -250,32 +250,18 @@ impl FailedPayableDao for FailedPayableDaoReal<'_> {
-                    )))
-                }
-            }
+            Ok(_rows) => Ok(()),
```
You could theoretically finish this branch the way you usually do. In terms of consistency, I'd prefer if you did that actually. 

How to proceed? Up there at the beginning of this method, you fetches all unchecked pending too long. That means you know how many of those there are. You can later check it against this returned number of updated 
rows ;) Those two numbers should match.

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -51,20 +50,17 @@ pub struct FailedTx {
                     f,
                     "WHERE reason = 'PendingTooLong' AND rechecked = 0 \
-                     AND timestamp >= {} \
                      ORDER BY timestamp DESC",
```
I bet we don't even need to order the records now. It's of no use. We will treat them all the same. 

Feel free to delete that condition. It shouldn't kill your tests.

bertllll commented on node/src/accountant/db_access_objects/failed_payable_dao.rs:
```diff
@@ -578,50 +592,34 @@ mod tests {
-        let result = subject.retrieve_txs(Some(FailureRetrieveCondition::UncheckedPendingTooLong(
-            3600,
-        )));
+        let result = subject.retrieve_txs(Some(FailureRetrieveCondition::UncheckedPendingTooLong));
```
Actually, this should be the act of this test. Don't isolate `insert_new_records`. (It's the fake prophet :laughing: )

bertllll committed:
GH-602: Scanner system revamp (#640)

* GH-602: big bunch of changes

* GH-602: interim commit

* GH-602: four scans layout installed

* GH-602: mostly done but still some crumbs of work left

* GH-602: interim commit

* GH-602: Recorder enriched by counter messages

* GH-602: removed obfuscated and made right the formerly developed StopConditions

* GH-602: removed obfuscated and made right the formerly developed StopConditions

* GH-602: main part probably done

* GH-602: more todos knocked off in accountant/mod.rs

* GH-602: deep dive into schedulers and the awarness of the schedule state

* GH-602: save point before going into fully private scanner

* GH-602: another save point, this time before turning the scanner into fully private object

* GH-602: finally fully scanners as flexible parametrized generic objects

* GH-602: todos placed (many) as understood as the whole design

* GH-602: more cases covered...knocking off todos!() continues

* GH-602: made a markup for upcoming steps

* GH-602: made a markup for upcoming steps

* GH-602: Manual scanners prevented by the automation. Tested

* GH-602: More tests written; mainly constraining the manual scanning

* GH-602: before trying to implement the ScanScheduleHintErrorResolver

* GH-602: tests for pending payable scanner and its start scan error resolver are compiling

* GH-602: all tests in scan_schedulers passing now

* GH-602: all tests in scan_schedulers passing now

* GH-602: slowly fixing test in accountant/mod.rs

* GH-602: fixed another couple of failing tests in accountant/mod.rs and scanners/mod.rs

* GH-602: adding another little pile

* GH-602: closing the gap... probably only one test to be written yet

* GH-602: one more test written...ext tr. pen payable

* GH-602: eliminated last todos in scanners/mod.rs

* GH-602: first rough clean-up

* GH-602: some comments fixed

* GH-602: first piece of the review finished

* GH-602: addressing review continues...

* GH-602: addressing review continues on and on...

* GH-602: more things for the review...

* GH-602: last 6 comments to address

* GH-602: finished the big two tests covering the very start of the automatic process with pp scanner playing the first violin

* GH-602: huge tests refactored

* GH-602: small functionalities made right and huge tests refactored

* GH-602: prepared some fixes for early rescheduling

* GH-602: moved the scanner test utils

* GH-602: counter message summary comment added

* GH-602: probably finished the minimm of what can be submitted seriously

* GH-602: removed commented out impl

* GH-602: vast majority of the review is covered

* GH-602: refactoring in the biggeste tests - a la utkarsh

* GH-602: review two ends

* GH-602: improved comments

---------

Co-authored-by: Bert <Bert@Bert.com>

Commit:
Merge branch 'master' into GH-598

utkarshg6 committed:
GH-631: add more review changes

utkarshg6 committed:
GH-631: introduce the replace_record() in SentPayableDao

utkarshg6 committed:
v0.9.0: version bump (#657)

utkarshg6 committed:
GH-631: review 4 changes

utkarshg6 committed:
GH-631: review 5 changes

utkarshg6 committed:
GH-631: review 6 changes

utkarshg6 committed:
GH-631: import the Itertools and make join work without collect

utkarshg6 committed:
Merge branch 'master' into GH-631

utkarshg6 committed:
Merge branch 'GH-598' into GH-631
